tests/test_shell_launcher.py

```diff
@@ -0,0 +1,323 @@
+        mock_open.assert_called_once()
+
+
+def test_shell_launcher_start_calls_popen_with_value(
+    shell_launcher, shell_cmd: ShellLauncherCommand
+):
+    """Test that popen was called with correct values"""
+    with unittest.mock.patch(
+        "smartsim._core.shell.shellLauncher.sp.Popen"
+    ) as mock_open:
+        _ = shell_launcher.start(shell_cmd)
+        mock_open.assert_called_once_with(
+            shell_cmd.command_tuple,
+            cwd=shell_cmd.path,
+            env=shell_cmd.env,
+            stdout=shell_cmd.stdout,
+            stderr=shell_cmd.stderr,
+        )
+
+
+def test_popen_returns_popen_object(shell_launcher, test_dir: str):
+    """Test that the popen call returns a popen object"""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    with (
+        open(out_file, "w", encoding="utf-8") as out,
+        open(err_file, "w", encoding="utf-8") as err,
+    ):
+        cmd = ShellLauncherCommand(
+            {},
+            run_dir,
+            subprocess.DEVNULL,
+            subprocess.DEVNULL,
+            EchoHelloWorldEntity().as_program_arguments(),
+        )
+        id = shell_launcher.start(cmd)
+    with shell_launcher._launched[id] as proc:
+        assert isinstance(proc, sp.Popen)
+
+
+def test_popen_writes_to_output_file(shell_launcher, test_dir: str):
+    """Test that popen writes to .out file upon successful process call"""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    with (
+        open(out_file, "w", encoding="utf-8") as out,
+        open(err_file, "w", encoding="utf-8") as err,
+    ):
+        cmd = ShellLauncherCommand(
+            {}, run_dir, out, err, EchoHelloWorldEntity().as_program_arguments()
+        )
+        id = shell_launcher.start(cmd)
+        out.close(), err.close()
+    proc = shell_launcher._launched[id]
+    assert proc.wait() == 0
+    assert proc.returncode == 0
+    with open(out_file, "r", encoding="utf-8") as out:
+        assert out.read() == "Hello World!\n"
+    with open(err_file, "r", encoding="utf-8") as err:
+        assert err.read() == ""
+
+
+@requires_slurm
+def test_popen_fails_with_invalid_cmd(shell_launcher, test_dir):
+    """Test that popen returns a non zero returncode after failure"""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    with (
+        open(out_file, "w", encoding="utf-8") as out,
+        open(err_file, "w", encoding="utf-8") as err,
+    ):
+        args = (helpers.expand_exe_path("srun"), "--flag_dne")
+        cmd = ShellLauncherCommand({}, run_dir, out, err, args)
+        id = shell_launcher.start(cmd)
+        proc = shell_launcher._launched[id]
+        proc.wait()
+        assert proc.returncode != 0
+        with open(out_file, "r", encoding="utf-8") as out:
+            assert out.read() == ""
+        with open(err_file, "r", encoding="utf-8") as err:
+            content = err.read()
+            assert "unrecognized option" in content
```

Comment:
Not wrong just curious: Does this strictly need to use `srun`? I looks like this test is just testing the presence of an unrecognized flag. Could this just use `ls` or some other native tool we can expect to find on any *nix machine?

Reply:
oh yes good call!